pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,279 @@
+		return framework.NewStatus(framework.Wait, ""), 10 * PermitWaitingTime
+	}
+
+	klog.V(3).Infof("The count of podGroup %v is up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+		podGroupName, minAvailable, running, waiting)
+	cs.FrameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		if waitingPod.GetPod().Namespace == namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
+			klog.V(3).Infof("Permit allows the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
+			waitingPod.Allow(cs.Name())
+		}
+	})
+
+	delete(cs.PodGroupInfos, fmt.Sprintf("%v/%v", namespace, podGroupName))
+	return framework.NewStatus(framework.Success, ""), 0
+}
+
+// Unreserve is called by the scheduling framework when a reserved pod was
+// rejected in a later phase.
+func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) {
+	podGroupName, exist := p.Labels[PodGroupName]
+	if !exist {
+		return
+	}
+
+	cs.FrameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		if waitingPod.GetPod().Namespace == p.Namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
+			klog.V(3).Infof("Unreserve rejects the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
+			waitingPod.Reject(cs.Name())
+		}
+	})
+}
+
+// GetPodGroupLabels will check the pod if belongs to some podGroup. If so, it will return the
+// podGroupName、minAvailable of podGroup. If not, it will return "" as podGroupName.
+func GetPodGroupLabels(p *v1.Pod) (string, int) {
+	podGroupName, exist := p.Labels[PodGroupName]
+	if !exist {
+		return "", 0
+	}
+	minAvailable, exist := p.Labels[PodGroupMinAvailable]
+	if !exist {
+		return "", 0
+	}
+	minNum, err := strconv.Atoi(minAvailable)
+	if err != nil {
+		klog.Error(err)
+		return "", 0
+	}
+	return podGroupName, minNum
+}
+
+func (cs *Coscheduling) calculateTotalPods(podGroupName, namespace string) int {
+	// TODO remove the hack manner to get the total pods
+	selector := labels.Set{PodGroupName: podGroupName}.AsSelector()
+	pods, err := cs.PodLister.Pods(namespace).List(selector)
+	if err != nil {
+		klog.Error(err)
+		return 0
+	}
+	return len(pods)
+}
+
+func (cs *Coscheduling) calculateAllocatedPods(podGroupName, namespace string) int {
+	pods, err := cs.FrameworkHandle.SnapshotSharedLister().Pods().FilteredList(func(pod *v1.Pod) bool {
+		if pod.Labels[PodGroupName] == podGroupName && pod.Namespace == namespace {
+			return true
+		}
+		return false
+	}, labels.NewSelector())
+
+	if err != nil {
+		klog.Error(err)
+		return 0
+	}
+
+	return len(pods)
+}
+
+func (cs *Coscheduling) calculateRunningPods(podGroupName, namespace string) int {
+	running := 0
+	_, err := cs.FrameworkHandle.SnapshotSharedLister().Pods().FilteredList(func(pod *v1.Pod) bool {
+		if pod.Labels[PodGroupName] == podGroupName && pod.Namespace == namespace && pod.Status.Phase == v1.PodRunning {
```

Comment:
It will take some time for Pod to reach Running phase after scheduling. Assuming the MinAvailable is 2, and two Pods pass the Permit and start binding, if the 3rd Pod of the same group arrives, will this Pod be put into wait list and eventually timeout if it is the last Pod in the PodGroup?

Reply:
Yes, it may lead to the 3rd Pod timeout. It's a tougher condition to Coscheduling. But the 3rd Pod timeout will not effect the job because the min pods is created.

Reply:
What about using PodInfoGroup? Also, do you actually need to check they are running or just that they are scheduled? If the latter, you don't need to check the Phase. Only scheduled pods are visible in the snapshot.

Reply:
In this place, I want to check if they are running instead of scheduled because 
1. The Unreserve and Permit is Asynchronous. So maybe some pods is timeout in Permit and before the Unreserve. They are scheduled but can be used to calculate the total count in Permit.
2. the pod belongs to the PodGroup may be created failed or pulling image failed..... This situation still causes the whole job to wait for a long time because of less than min.

Reply:
> Or perhaps we can expose this information (if its assumed or actually scheduled) in the SharedLister.
SG. Maybe I can use the assumed here in this version 1.18 here and expose the status(assumed/actually) in framework in 1.19. WDYT? @alculquicondor

Reply:
Did you mean the binding?

Reply:
This should be part of the README, at least.